crates/bevy_ecs/hecs/src/dynamic_query.rs

```diff
@@ -0,0 +1,418 @@
+    dynamic_components: Arc<Mutex<HashSet<DynamicComponentInfo>>>,
+}
+
+impl DynamicComponentInfoRegistry {
+    /// Get the component info for the given Rust type
+    pub fn get_rust_component_info<T: 'static>(&self) -> DynamicComponentInfo {
+        DynamicComponentInfo {
+            id: TypeId::of::<T>().into(),
+            layout: Layout::from_size_align(size_of::<T>(), align_of::<T>()).unwrap(),
+            drop: None,
+        }
+    }
+
+    /// Registers a dynamic component and returns [`DynamicComponentInfo`] which can be used to
+    /// spawn dynamic components with an [`EntityBuilder`].
+    ///
+    /// Returns `None` if a component with the provided ID has already been registered.
+    pub fn register_dynamic_component(
+        &self,
+        external_id: u64,
+        layout: Layout,
+        drop: Option<fn(*mut u8)>,
+    ) -> Option<DynamicComponentInfo> {
+        let info = DynamicComponentInfo {
+            id: ComponentId::ExternalId(external_id),
+            layout,
+            drop,
+        };
+
+        if self.dynamic_components.lock().unwrap().insert(info) {
+            Some(info)
+        } else {
+            None
+        }
+    }
+
+    /// Return the component info for a non-Rust dynamic component. If the dynamic component has not
+    /// been registered yet, it will return None.
+    pub fn get_dynamic_component_info(&self, external_id: u64) -> Option<DynamicComponentInfo> {
+        self.dynamic_components
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|&x| x.id == ComponentId::ExternalId(external_id))
+            .map(Clone::clone)
+    }
+}
+
+lazy_static::lazy_static! {
+    /// The global registry of component info
+    pub static ref DYNAMIC_COMPONENT_INFO_REGISTRY: DynamicComponentInfoRegistry =
+        DynamicComponentInfoRegistry::default();
+}
+
+/// A Query that can be constructed at runtime
+#[derive(Default)]
+pub struct DynamicQuery {
+    /// Whether or not the entity should be queried
+    pub entity: bool,
+    // fields private and then checking the values in the constructor.
+    /// The list of accesses to immutable components
+    pub immutable: Vec<DynamicComponentInfo>,
+    /// The list of accesses to mutable components
+    pub mutable: Vec<DynamicComponentInfo>,
+}
+
+impl DynamicQuery {
+    /// Returns how, if at all, the query accesses the given archetype
+    pub fn access(&self, archetype: &Archetype) -> Option<Access> {
+        let mut access = None;
+
+        for component in self.immutable.iter() {
+            if archetype.has_component(component.id) {
+                access = Some(Access::Read);
+            }
+        }
+
+        for component in self.mutable.iter() {
+            if archetype.has_component(component.id) {
+                access = Some(Access::Write);
+            }
+        }
+
+        access
+    }
+
+    /// Acquires a dynamic borrow to the archetype
+    // FIXME: Figure out if we need this
```

Comment:
I'm a little confused on whether or not the query implementation needs to borrow from the archetype. It seems like `hecs`'s query, which I modelled `DynamicQuery` off of, doesn't borrow the Archetype, but bevy's Query does. I'll probably need a pointer on that one.

Reply:
But that's what confused me, because I copied the `DynamicQueryBorrow` implementation from the `hecs::QueryBorrow` implementation, but it never calls `Fetch::Borrow`. In fact I never found anywhere in `hecs` that called `Fetch::borrow`, but I could have missed it.

Oh, I wonder if `bevy_hecs` has been modified in this respect, though. Maybe I need to compare `bevy_hecs` to `hecs` to see what the differences are.

Reply:
OK, I think I've fixed the borrowing. Does this look right to you?

When I create the chunk iterator, which is going to be dishing out references to the archetype, I borrow the archetype:

https://github.com/bevyengine/bevy/blob/4a4c7fb2dfb5342836158c737ffeaa66ef665633/crates/bevy_ecs/hecs/src/dynamic_query.rs#L324-L335

Then when the `DynamicChunkIter` is dropped, we release the borrow:

https://github.com/bevyengine/bevy/blob/4a4c7fb2dfb5342836158c737ffeaa66ef665633/crates/bevy_ecs/hecs/src/dynamic_query.rs#L395-L399